### DISCUSSION TITLE
Google OAuth error: invalid grant

### DISCUSSION THREAD

**Discussion by Graciasjr** [pocketbase/pocketbase#2476] [Q&A] [answered]
Hello please I have an error to submit. When I actually try google oauth authentication I first get a 400 error from a POST method. And in metadata returns the following object

```
"errorMessage":"oauth2":cannot fetch token:400 Bad Request\nResponse:{\n \"error\":\"invalid_grant",\n\"error_description\":\"Invalid code vérifier.\"\n}
}
```
so I heard that this error is probably due to a time difference between the Google servers and the server that launches the request but I don't know if it's true.

**Comment by ganigeorgiev** [comment]
Did you follow the examples from https://pocketbase.io/docs/authentication/#oauth2-integration?

Could you provide a code sample how do you authenticate?

**Reply by Graciasjr** [reply]
yes i follow examples
i use sveltekit

in my hooks.server.js

```
import PocketBase from 'pocketbase';

export const handle = async({event,resolve}) =>{
    event.locals.pb= new PocketBase('http://127.0.0.1:3001');
    event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');

    try {        
            event.locals.pb.authStore.isValid && await event.locals.pb.collection('users').authRefresh();
        
        
    } catch (error) {
        event.locals.pb.authStore.clear();
    }

    response.headers.append('set-cookie',event.locals.pb.authStore.exportToCookie());

    return response;
}

```

in my +layout.server.js

```
xport const load = async({locals,url})=>{
    
    const authMethods = await locals.pb?.collection('users').listAuthMethods();
    if(!authMethods){
        return{
            authProviderRedirect:'',
            authProviderState:''
        };
    }

    const redirectURL =`${url.origin}/callback`;
    const googleAuthProvider = authMethods.authProviders[0];
    const authProviderRedirect = `${googleAuthProvider.authUrl}${redirectURL}`;
    const state = googleAuthProvider.state;
    
    return {
        authProviderRedirect:authProviderRedirect,
        authProviderState:state,
    };

```

in my layout.svelte

``` 
    function auth(){
            if(browser){
                document.cookie = `state=${data?.authProviderState}`;
            }
            window.location.href = data.authProviderRedirect || '';
        }

```

in my callback +server.js

```
    try {
        await locals.pb.collection('users').authWithOAuth2(provider.name,code || '',provider.codeVerifier,redirectURL);
        throw redirect(303,'/')

    } catch (error) {
        //The following code help me to see error message
        throw redirect(303,`/folders?error=${error}`)

    }
```

**Reply by ganigeorgiev** [reply]
The above code sample doesn't show how `provider.codeVerifier` is created.

Make sure that the `provider` object is the same as the `googleAuthProvider` object that is used to generate the OAuth2 auth url.

Note that calling `listAuthMethods()` will generate a new code verifier and challenge every time when you call it.

**Reply by Graciasjr** [reply]
```
    const authMethods = await locals.pb?.collection('users').listAuthMethods();
    const provider = authMethods.authProviders[0];

// provider.codeVerifier

```

**Reply by ganigeorgiev** [reply, marked as answer]
`listAuthMethods()` will generate a new `codeVerifier` that will differ from the challenge created in your `+layout.server.js`. You have to use the `codeVerifier` from that step, eg. pass it down similar to the `state` parameter.

**Reply by Graciasjr** [reply]
thank you so much 🙇🏿‍♂️ it works succesfully no more server side error pb but this time an error 500:internal error redirected to the callback route

**Reply by Graciasjr** [reply]
Please, is there a way to access the token returned by Google auth?

**Reply by ganigeorgiev** [reply]
Yes, it is available under the `meta.accessToken` prop in the returned OAuth2 auth response.

**Reply by Graciasjr** [reply]
is it the pb_auth cookie stored in my browser that is the response sent by Oauth? otherwise, I can't find the `meta` object

**Reply by ganigeorgiev** [reply]
The `meta` prop is returned with the success OAuth2 auth call:

```js
const authData = await locals.pb.collection('users').authWithOAuth2Code(...)

console.log(authData.meta)
```

**Reply by Graciasjr** [reply]
Thank you so much It's good. I had used the wrong process.
